DEVELOPMENT.md

```diff
@@ -0,0 +1,42 @@
+# Development Guidelines
+
+This document describes the development guidelines when writing and maintaining EnMasse source code. 
+
+## General considerations
+
+As EnMasse is a polyglot project with multiple languages in the same repo, and uses `make` as the
+single entry point for invoking a full project build. Make will invoke the language-specific build
+system.
+
+### Introducing new dependencies
+
+If you need a function or class utility that requires adding a new dependency, first consider:
+
+* If the dependency has a reliable release schedule and track record for security updates.
+* If the dependency pulls in transitive dependencies that increase the footprint significantly
+* If you can copy the function or class (assuming the license allows it) into the EnMasse source tree. This should only be done for trivial amounts of code where the maintenance cost is expected to be low.
+* If the cost of implementing and maintaining the functionality yourself is worthwhile in the same manner as above.
+
+## Java
+
+Java source code is built using Maven.
+
+Follow the [Oracle Java Code Conventions](https://www.oracle.com/technetwork/java/codeconvtoc-136057.html).
+
+Prefer to use plain Java whenever you can. Code should be easy to understand without intimate knowledge of specific frameworks. VertX, Jackson, JAX-RS and tests framework annotations such as Junit are in use at present, and should be preferred instead of introducing other frameworks. However, some dependencies like Eclipse Hono uses Spring, in which case it cannot be ignored.
+
```

Comment:
should we add a note about the java version we use ?

Reply:
Actually on a second thought @jbtrystram , i wonder if this is better stated in HACKING.md. Otherwise we will have to remember to update this document everytime we change javascript, java or go version, which doesn't really change anything around the guidelines.